gradle-junit-reports/src/main/java/com/palantir/gradle/junit/BuildJunitReportService.java

```diff
@@ -0,0 +1,108 @@
+import org.w3c.dom.Document;
+
+public abstract class BuildJunitReportService
+        implements BuildService<ReportFile>, OperationCompletionListener, AutoCloseable {
+    private static final Logger log = LoggerFactory.getLogger(BuildJunitReportService.class);
+    private final List<Report.TestCase> testCases = new ArrayList<>();
+    private final long startTimeNanos = System.nanoTime();
+
+    public static class ReportFile implements BuildServiceParameters, Serializable {
+        private File reportFile;
+
+        public final void setReportFile(File reportFile) {
+            this.reportFile = reportFile;
+        }
+    }
+
+    @Override
+    public final synchronized void onFinish(FinishEvent finishEvent) {
+        if (!(finishEvent instanceof TaskFinishEvent)) {
+            return;
+        }
+        TaskFinishEvent taskFinishEvent = (TaskFinishEvent) finishEvent;
+        TaskOperationResult result = taskFinishEvent.getResult();
+        String taskName = taskFinishEvent.getDescriptor().getTaskPath();
+        Report.TestCase.Builder testCase = new Report.TestCase.Builder().name(taskName);
+
+        if (result instanceof TaskFailureResult) {
+            TaskFailureResult failure = (TaskFailureResult) result;
+            List<? extends org.gradle.tooling.Failure> failures = failure.getFailures();
+            if (failures.isEmpty()) {
+                testCase.failure(new Report.Failure.Builder()
+                        .message("operation failed without any specific failure information")
+                        .details("")
+                        .build());
+            } else {
+                if (failures.size() != 1) {
+                    log.warn("task had multiple failures, which is unexpected. Reporting the first one");
+                }
+                org.gradle.tooling.Failure firstFailure = failures.iterator().next();
+                testCase.failure(new Report.Failure.Builder()
+                        .message(firstFailure.getMessage())
+                        .details(firstFailure.getDescription())
+                        .build());
+            }
+        }
+        testCases.add(testCase.build());
+    }
+
+    private Path getTargetFile() {
+        return getParameters().reportFile.toPath();
+    }
+
+    @Override
+    public final synchronized void close() throws Exception {
+        Report report = new Report.Builder()
+                .name("gradle")
+                .subname("gradle")
+                .elapsedTimeNanos(System.nanoTime() - startTimeNanos)
+                .addAllTestCases(testCases)
+                .build();
+        Document xml = JunitReportCreator.reportToXml(report);
+
+        Files.createDirectories(getTargetFile().getParent());
+
+        try (Writer writer = Files.newBufferedWriter(getTargetFile(), StandardCharsets.UTF_8)) {
+            XmlUtils.write(writer, xml);
+        }
+    }
+
+    public final List<Report.TestCase> getTestCases() {
```

Comment:
I think this is visible only for tests, but given current model should synchronize

```suggestion
    public final synchronized List<Report.TestCase> getTestCases() {
```